### PR TITLE
Implement Stdio Channel

### PR DESCRIPTION
**Pull request by schrieveslaach**
This PR adds support for serving the language over stdio channel which is the default communication channel in tools such as Neovim. That use case has been also reported in the forum (see [here](https://community.sonarsource.com/t/running-sonarlint-language-server-from-shell/24440/14)).

The changes include but are still work in progress: 

- Use picocli to improve CLI parsing
- Start language server via stdio channeling
- Avoid NPE if client doesn't send initialize params

The language server works with Neovim (started with `java -jar /path/to/sonarlint-language-server/target/sonarlint-language-server-2.14.0-SNAPSHOT.jar -stdio -analyzers ~/Downloads/sonar/extension/analyzers/sonarjava.jar` and shown by `:LspInfo`, shown in the screencast) but there are no reports.

output

I have some lua stubs in my Neovim config in place but I'm not sure what is missing to get reports.

```lua
local M = {}

M.client_id = nil

function M.start_or_attach()
   local bufnr = vim.api.nvim_get_current_buf()

   local root_dir = vim.fs.dirname(vim.fs.find({'.git'}, { upward = true })[1])

   local config = {}
   config.name = 'sonarlint'
   config.cmd = { '/home/marc/Projekte/Open-Source/sonarlint-language-server/bin/sonarlint' }
   config.root_dir = root_dir
   config.handlers = {}
   config.handlers['sonarlint/isOpenInEditor'] = function(...)
      vim.pretty_print(...)
      return true
   end
   config.handlers['sonarlint/isIgnoredByScm'] = function(...)
      vim.pretty_print(...)
      return false
   end
   config.handlers['sonarlint/showSonarLintOutput'] = function(...)
      vim.pretty_print(...)
      return false
   end
   config.handlers['sonarlint/showSonarLintOutput'] = function(...)
      vim.pretty_print(...)
      return false
   end

   if not M.client_id then
      M.client_id = vim.lsp.start_client(config)
   end

   vim.lsp.buf_attach_client(bufnr, M.client_id)
end

return M
```

Any feedback is welcome.

### CODE REVIEW

**Comment by henryju** [comment]
Hi, SonarJava is a complex analyzer, that requires extra requests not part of the LSP (in order to get the classpath of the file to analyze). Maybe you should test with a simpler analyzer like SonarPHP or SonarPython first.

**Comment by schrieveslaach** [comment]
@henryju, thanks for the hint. Now, I tried it with just the SonarPython (cmd option: ` -analyzers ~/Downloads/sonar/extension/analyzers/sonarpython.jar`) and I requested `sonarlint/listAllRules`. However, the LSP replies with an internal error:

```
Jan. 18, 2023 7:23:46 VORM. org.eclipse.lsp4j.jsonrpc.RemoteEndpoint fallbackResponseError\nSCHWERWIEGEND: Internal error: java.lang.IllegalStateException: java.lang.IllegalStateException: Unable to get settings in time\njava.util.concurrent.CompletionException: java.lang.IllegalStateException: java.lang.IllegalStateException: Unable to get settings in time\n\tat java.base/java.util.concurrent.CompletableFuture.encodeThrowable(CompletableFuture.java:314)\n\tat java.base/java.util.concurrent.CompletableFuture.completeThrowable(CompletableFuture.java:319)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:645)\n\tat java.base/java.util.concurrent.CompletableFuture$Completion.exec(CompletableFuture.java:479)\n\tat java.base/java.util.concurrent.ForkJoinTask.doExec(ForkJoinTask.java:290)\n\tat java.base/java.util.concurrent.ForkJoinPool$WorkQueue.topLevelExec(ForkJoinPool.java:1020)\n\tat java.base/java.util.concurrent.ForkJoinPool.scan(ForkJoinPool.java:1656)\n\tat java.base/java.util.concurrent.ForkJoinPool.runWorker(ForkJoinPool.java:1594)\n\tat java.base/java.util.concurrent.ForkJoinWorkerThread.run(ForkJoinWorkerThread.java:183)\nCaused by: java.lang.IllegalStateException: java.lang.IllegalStateException: Unable to get settings in time\n\tat org.sonarsource.sonarlint.ls.EnginesFactory.createStandaloneEngine(EnginesFactory.java:106)\n\tat org.sonarsource.sonarlint.ls.standalone.StandaloneEngineManager.getOrCreateStandaloneEngine(StandaloneEngineManager.java:37)\n\tat org.sonarsource.sonarlint.ls.CommandManager.listAllStandaloneRules(CommandManager.java:232)\n\tat org.sonarsource.sonarlint.ls.SonarLintLanguageServer.lambda$listAllRules$8(SonarLintLanguageServer.java:420)\n\tat java.base/java.util.concurrent.CompletableFuture$UniApply.tryFire(CompletableFuture.java:642)\n\t... 6 more\nCaused by: java.lang.IllegalStateException: Unable to get settings in time\n\tat org.sonarsource.sonarlint.ls.settings.SettingsManager.getCurrentSettings(SettingsManager.java:129)\n\tat org.sonarsource.sonarlint.ls.NodeJsRuntime.init(NodeJsRuntime.java:50)\n\tat org.sonarsource.sonarlint.ls.NodeJsRuntime.getNodeJsPath(NodeJsRuntime.java:70)\n\tat org.sonarsource.sonarlint.ls.EnginesFactory.createStandaloneEngine(EnginesFactory.java:94)\n\t... 10 more\n\n
```

Could you help me to understand what I need to do send to the LSP?

**Comment by henryju** [comment]
Our LSP implementation is at the moment very tight with VSCode. The current VSCode behavior is to send a `workspace/didChangeConfiguration` notification quickly after LS initialization, in order to provide current settings. We are relying on this notification to be delivered in a timely manner.

**Comment by schrieveslaach** [comment]
Thanks for the information and trying to guide me.  It should be easy to sent a notification from Lua/Neovim to the LSP (see example).

```lua
local client = vim.lsp.get_client_by_id(M.client_id)
client.notify("workspace/didChangeConfiguration", {
   -- I have no clue what Sonarlint expects me to send
   settings = {}
})
```

However, no idea what to sent in the notification. Do you could give me a hint where I find a payload example I'll have to sent when using the Python analyzer?

**Comment by henryju** [comment]
As far as I remember, you can send empty settings.

**Comment by schrieveslaach** [comment]
Got it working within Neovim (see screencast below) and I think this PR is ready to review.

output

File: pom.xml [hunk 1]
```diff
@@ -44,7 +44,7 @@
     <dependency>
       <groupId>org.eclipse.lsp4j</groupId>
       <artifactId>org.eclipse.lsp4j</artifactId>
-      <version>0.14.0</version>
+      <version>0.19.0</version>
```
Comment (henryju) [inline pom.xml:47]: Is this change really needed in this PR?

File: src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java [hunk 2]
```diff
@@ -19,114 +19,96 @@
...
+public class ServerMain implements Callable<Integer> {
 
   private final PrintStream out;
-  private final PrintStream err;
 
-  public ServerMain(PrintStream out, PrintStream err) {
+  @Parameters(index = "0", description = "The port to which sonarlint should connect to.", defaultValue = "-1")
+  private int deprecatedJsonRpcPort;
+
+  @Option(names = "-jsonRpcPort", description = "The port to which sonarlint should connect to.")
```
Comment (henryju) [inline src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java:47]: I don't think the name `jsonRpcPort` is good. Even when using stdio, the protocol is still JSON RPC. Maybe `-socketPort`, or `-clientPort` like in the JDT LS, or simply `-port`?

File: src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java [hunk 3]
```diff
@@ -19,114 +19,96 @@
...
+    validate();
 
-  Collection<Path> extractAnalyzers(String[] args) {
-    var indexOfAnalyzersParam = List.of(args).indexOf(ANALYZERS_KEY);
-    if (indexOfAnalyzersParam == -1) {
-      err.println(USAGE);
-      exitWithError();
+    if (useStdio) {
+      SonarLintLanguageServer.byStdio(analyzers, extraAnalyzers);
+      synchronized(this) {
```
Comment (henryju) [inline src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java:80]: What is this used for?

File: src/test/java/org/sonarsource/sonarlint/ls/ServerMainTests.java [hunk 4]
```diff
@@ -20,121 +20,76 @@
...
-    assertThat(paths).isEmpty();
-  }
-
-  @Test
-  void testExtractingExtraAnalyzersExitsOnMalformedPaths() {
-    var args = new String[] {"-extraAnalyzers", INVALID_PATH};
-
-    assertThrows(RuntimeException.class, () -> underTest.extractExtraAnalyzers(args));
-  }
+   private static final String INVALID_PATH = "\0invalid?;path\n";
```
Comment (henryju) [inline src/test/java/org/sonarsource/sonarlint/ls/ServerMainTests.java:35]: Formatting issue. SonarSource coding style is to use 2 spaces for indentation

Comment (schrieveslaach) [reply src/test/java/org/sonarsource/sonarlint/ls/ServerMainTests.java]: I applied the Eclipse setting in [eclipes.jdt.ls](https://github.com/eclipse/eclipse.jdt.ls) and formatted all touched files. But it rearranged some static imports. Is it intended to have the static imports at the top?

File: src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java (hunk 3 above)
Comment (schrieveslaach) [reply src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java]: That was kind of a hack and I'm open for suggestions. It is necessary to keep reading from STDIN open and if I remove it the process will stop immediately.

Maybe, that piece of code should be done in `SonarLintLanguageServer` and the object should be notified when the server receives a shutdown request.

**Comment by schrieveslaach** [comment]
> Overall, the change looks good to me. There are some formatting issues to solve (SonarSource coding style is available in this repo: https://github.com/SonarSource/sonar-developer-toolset I would create separate independent commits for more clarity:
> 
>     * move CLI argument parsing to picocli
> 
>     * add support for stdio
> 
>     * changes `initialize` to not have any mandatory parameter
> 
>     * update lsp4j

Will do the separate commits as soon as we figured out the remaining questions.

I'm not sure why the GitHub actions fail. Maybe, you can give me hint.

File: src/test/java/org/sonarsource/sonarlint/ls/ServerMainTests.java (hunk 4 above)
Comment (henryju) [reply src/test/java/org/sonarsource/sonarlint/ls/ServerMainTests.java]: Are you using https://github.com/SonarSource/sonar-developer-toolset/blob/master/eclipse/sonar.importorder? Static imports should all be after other imports.

File: src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java (hunk 3 above)
Comment (henryju) [reply src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java]: I suggest you look at how it is implemented in other language servers based on LSP4J, like https://github.com/eclipse/eclipse.jdt.ls. I'm pretty sure they are using stdio as well.

Comment (schrieveslaach) [reply src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java]: I found a solution for it by adding a `CountDownLatch` in `SonarLintLanguageServer` and waiting for it in `ServerMain`.

**Comment by schrieveslaach** [comment]
@henryju, I tried to rebase my PR but I noticed that `org.sonarsource.sonarlint.core:sonarlint-core:jar:8.14.0.61607 ` is not publicly available:

image

Anyway, it would be great to get some new feedback.

File: src/test/java/org/sonarsource/sonarlint/ls/mediumtests/AbstractLanguageServerMediumTests.java [hunk 5]
```diff
@@ -117,7 +121,7 @@ static void startServer() throws Exception {
 
     client = new FakeLanguageClient();
 
-    var executor = Executors.newSingleThreadExecutor();
+    var executor = Executors.newFixedThreadPool(2);
```
Comment (henryju) [inline src/test/java/org/sonarsource/sonarlint/ls/mediumtests/AbstractLanguageServerMediumTests.java:124]: What is the reason for this change?

**Review by henryju — COMMENTED** [review]
Overall LGTM except for a doubt on the executor service. @jblievremont may I let you handle this contribution, and see if it makes sense to merge it?

**Comment by jblievremont** [comment]
Hey 👋🏻 thank you for your contribution! I especially appreciate the introduction of `picocli`.

I rebased the branch on the current `master` and did some manual testing, there does not seem to be any impact in VSCode as far as I can tell.

I am not familiar with Neovim so I could not really check whether the `stdio` channel works.

What I can do is to open a duplicate pull request with my rebased branch, let our CI run on it and send you a link to the PR build - this should allow you to check the results in Neovim.

How does this sound to you?

File: src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java (hunk 2 above)
Comment (schrieveslaach) [reply src/main/java/org/sonarsource/sonarlint/ls/ServerMain.java]: I chose port `-port`.

File: src/test/java/org/sonarsource/sonarlint/ls/mediumtests/AbstractLanguageServerMediumTests.java (hunk 5 above)
Comment (schrieveslaach) [reply src/test/java/org/sonarsource/sonarlint/ls/mediumtests/AbstractLanguageServerMediumTests.java:124]: `ServerMain` will wait for a shutdown and therefore, I needed a second thread that starts the main, see `executor.submit(() -> cmd.execute(clonedArgs));` down below. If the test code here uses `newSingleThreadExecutor()`, the `clientSideLauncher` and `cmd` are dead locked.

That is necessary due to the adjustments in `ServerMain` and `SonarLintLanguageServer`. The adjusted structure of `ServerMain` and `SonarLintLanguageServer` should also provide the ability to fix the potential resource leak:

grafik

**Comment by schrieveslaach** [comment]
> Hey 👋🏻 thank you for your contribution! I especially appreciate the introduction of `picocli`.

I'm happy that this contribution is considered to be merged.

> 
> I rebased the branch on the current `master` and did some manual testing, there does not seem to be any impact in VSCode as far as I can tell.

Very nice. :sunglasses: 

> 
> I am not familiar with Neovim so I could not really check whether the `stdio` channel works.

I tested it locally and it works with a new Neovim plugin (see [sonarlint.nvim](https://gitlab.com/schrieveslaach/sonarlint.nvim)). I'm considering to test it there with a GitLab CI pipeline.

> 
> What I can do is to open a duplicate pull request with my rebased branch, let our CI run on it and send you a link to the PR build - this should allow you to check the results in Neovim.
> 
> How does this sound to you?

Actually, I'm a bit confused. The only thing that blocks me from doing the rebase myself, is the missing artifact `org.sonarsource.sonarlint.core:sonarlint-core:jar:8.14.0.61607` in Maven Central. Why is that? Previous versions are available in Maven central.

So my plan was to rebase and then split the commits as requested into separate ones ([see comment above](https://github.com/SonarSource/sonarlint-language-server/pull/187#pullrequestreview-1266998779)):

>  I would create separate independent commits for more clarity:

So I'm not sure if the duplicate pull request makes things easier or harder.

What's your impression here, @jblievremont?

**Comment by jblievremont** [comment]
Unfortunately, version 8.14.0.61607 of sonarlint-core is not available on Maven Central because we are in the middle of a sprint in which it has been quite heavily modified. This version is only accessible in the Maven repository that SonarSource developers use internally, and the stable/GA version of sonarlint-core 8.14.0 should be published soon to Maven Central.

The idea I had behind the duplicate PR is that since our CI is allowed to access the internal Maven repository, we would be able to rebase on the current master and let the checks run. I had completely missed @henryju's comments above 🙈.

We are quite aware - and sorry 😅 - that this makes it harder for people from the community to contribute.

What I suggest is to wait a couple of weeks for the GA releases of both sonarlint-core and sonarlint-language-server: it should make it easier for you to rebase your PR and split the commits.

**Comment by schrieveslaach** [comment]
Got it.

I found a workaround for doing it my self. I checked out https://github.com/SonarSource/sonarlint-core.git, used `mvn versions:set -DnewVersion=8.14.0.62524`, ran `mvn install` and then I could rebase and test it locally.

I will do the commit splitting now.

**Comment by schrieveslaach** [comment]
Rebased and splitted commits. @jblievremont, @henryju, I think the PR is ready to be merged.

**Comment by schrieveslaach** [comment]
I noticed that after rebasing I'm unable to start sonarlint-ls via the CLI:

```
❯ java -jar target/sonarlint-language-server-2.15.0-SNAPSHOT.jar -stdio -analyzers target/plugins/sonarpython.jar
java.lang.IllegalStateException: Embedded plugin not found: HTML
	at org.sonarsource.sonarlint.ls.SonarLintLanguageServer.lambda$addPluginPathOrFail$16(SonarLintLanguageServer.java:672)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at org.sonarsource.sonarlint.ls.SonarLintLanguageServer.addPluginPathOrFail(SonarLintLanguageServer.java:672)
	at org.sonarsource.sonarlint.ls.SonarLintLanguageServer.getEmbeddedPluginsToPath(SonarLintLanguageServer.java:663)
	at org.sonarsource.sonarlint.ls.SonarLintLanguageServer.<init>(SonarLintLanguageServer.java:216)
	at org.sonarsource.sonarlint.ls.SonarLintLanguageServer.byStdio(SonarLintLanguageServer.java:265)
	at org.sonarsource.sonarlint.ls.ServerMain.call(ServerMain.java:68)
	at org.sonarsource.sonarlint.ls.ServerMain.call(ServerMain.java:37)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1953)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2358)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2314)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine$RunLast.execute(CommandLine.java:2316)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at org.sonarsource.sonarlint.ls.ServerMain.main(ServerMain.java:98)
```

I just want to use the python analyzer.

It that happening due to my workaround (see above) or is it due to other changes within sonarlint-ls?

**Comment by schrieveslaach** [comment]
I noticed that #191 added the following code block to `SonarLintLanguageServer`:

```java
    addPluginPathOrFail("html", Language.HTML, plugins);
    addPluginPathOrFail("js", Language.JS, plugins);
    addPluginPathOrFail("xml", Language.XML, plugins);
    addPluginPathOrFail("text", Language.SECRETS, plugins);
```

Only if I provide following arguments, I'm able to start the CLI project: `java -jar target/sonarlint-language-server-2.16.0-SNAPSHOT.jar -stdio -analyzers target/plugins/sonarpython.jar target/plugins/sonarhtml.jar target/plugins/sonarjs.jar target/plugins/sonarxml.jar target/plugins/sonartext.jar`

@henryju, @jblievremont, is that intended? 

BTW: rebased with current master and PR works with the applied workaround locally.

**Comment by schrieveslaach** [comment]
@henryju, @jblievremont, is there something left for me before accepting this PR?

**Comment by jblievremont** [comment]
@schrieveslaach I think we will merge your contribution today (thanks again!), and I created a [ticket](https://sonarsource.atlassian.net/browse/SLLS-133) to address your previous comment about the "required" analyzers.

**Comment by schrieveslaach** [comment]
Awesome, thanks for accepting my contribution.

**Comment by albertocerrone** [comment]
Hey @schrieveslaach! Thank you for your contribution.
So now is it possible to use Sonarlint with Neovim? How do I do that?
Thank you so much ❤️

**Comment by schrieveslaach** [comment]
> Hey @schrieveslaach! Thank you for your contribution. So, is it possible now to use Sonarlint with Neovim? If yes how do I do that? Thank you so much heart

@albertocerrone, happy to provide value in that regard. 

I'm experimenting with a new Neovim plugin ([sonarlint.nvim](https://gitlab.com/schrieveslaach/sonarlint.nvim)). It is a very early and Java isn't working yet but I think I know what to do to get that running. But due to life my progress is very slow. Sorry for that.

**Comment by albertocerrone** [comment]
Thank you so much @schrieveslaach! You rock🔥

**Comment by gnujeremie** [comment]
Thanks @schrieveslaach , we switched recently at work to sonarcloud, your contributions are perfectly timed for me :) 👍

**Comment by schrieveslaach** [comment]
FYI: I hope that by the end of the year I'm able to see in Neovim that I'm using inefficient Java collections to reduce energy footprint of application.

https://fosstodon.org/@schrieveslaach/109683343437764941

**Comment by jblievremont** [comment]
For your information, sonarlint-language-server 2.16.0.65434 has been released and [published to Maven Central](https://repo1.maven.org/maven2/org/sonarsource/sonarlint/ls/sonarlint-language-server/2.16.0.65434/) 🚀 

We also started to use the stdio channel in SonarLint for VSCode, it seems that it improved performance and stability in some environments.

Thank you again for this great contribution!

**Comment by schrieveslaach** [comment]
@jblievremont, FYI: I updated my new AUR package sonarlint-ls (see [here](https://gitlab.com/schrieveslaach/aur-sonarlint-ls/-/commit/ce5d991a4fb097a6f24e445ebc3b0e0f935b56dc)) but to apply a patch because the property `sonar.python.version` does not point to a version that is available in Maven central.
